// @sap/approuter/lib/configuration.js
'use strict';

const _ = require('lodash');

const applicationConfig = require('./configuration/app-config');
const configurationUtils = require('./utils/configuration-utils');
const environment = require('./environment');
const environmentConfig = require('./configuration/env-config');
const uaaConfiguration = require('./configuration/uaa-config');
const iasConfiguration = require('./configuration/ias-config');
const validators = require('./configuration/validators');

module.exports.load = function(options) {
  let workingDir = environment.getWorkingDirectory(options.workingDir, options.xsappConfig);

  let envConfig = environmentConfig.load(workingDir);
  let uaaConfig = uaaConfiguration.load(workingDir);
  let iasConfig = iasConfiguration.load(workingDir);

  let routerConfig = {
    appConfig: applicationConfig.loadConfiguration(workingDir, options.xsappConfig || 'xs-app.json',
      envConfig.destinations, uaaConfig.xsappname),
    serverPort: environment.getPort(options.port),
    workingDir: workingDir,
    getToken: options.getToken,
    getRouterConfig: options.getRouterConfig,
    uaaConfig: {
      options: uaaConfig
    },
    iasConfig: {
      options: iasConfig
    },
    httpsOptions: options.httpsOptions
  };

  return exports.createRouterConfig(routerConfig, envConfig);
};

module.exports.createRouterConfig = function(routerConfig, envConfig) {
  mergeConfiguration(routerConfig, envConfig);
  configurationUtils.mergePluginsIntoRoutes(routerConfig.appConfig, routerConfig.plugins);
  // should be after plugins are merged with xs-app.json provided routes
  validateUaaConfiguration(routerConfig);
  validateIasConfiguration(routerConfig);
  return routerConfig;
};

function mergeConfiguration(routerConfig, envConfig) {
  routerConfig.sessionTimeout = envConfig.sessionTimeout || routerConfig.appConfig.sessionTimeout;
  delete routerConfig.appConfig.sessionTimeout;

  routerConfig.appConfig.compression = routerConfig.appConfig.compression || {};
  _.merge(routerConfig.appConfig.compression, envConfig.compression);

  routerConfig.uaaConfig.tenantHostPattern = envConfig.tenantHostPattern;
  routerConfig.iasConfig.tenantHostPattern = envConfig.tenantHostPattern;

  envConfig = _.omit(envConfig, ['sessionTimeout', 'compression', 'tenantHostPattern']);
  routerConfig = _.merge(routerConfig, envConfig);

  return routerConfig;
}

function validateUaaConfiguration(routerConfig) {
  if (!isUaaCheckRequired(routerConfig.appConfig)) {
    return;
  }

  let uaaConfig = routerConfig.uaaConfig;
  if (Object.keys(uaaConfig.options).length === 0) {
    throw new Error('No UAA service found');
  }
  validators.validateUaaOptions(uaaConfig.options);

  if (uaaConfig.options.tenantmode === 'shared' && !uaaConfig.tenantHostPattern) {
    throw new Error('UAA tenant mode is shared, but environment variable TENANT_HOST_PATTERN is not set');
  }
}

function isUaaCheckRequired(appConfig) {
  return appConfig.authenticationMethod !== 'none' &&
         appConfig.routes.some(function(route) {
           return route.authenticationType === 'xsuaa';
         });
}

function validateIasConfiguration(routerConfig) {
  if (!isIasCheckRequired(routerConfig.appConfig)) {
    return;
  }

  let iasConfig = routerConfig.iasConfig;
  if (Object.keys(iasConfig.options).length === 0) {
    throw new Error('No IAS service found');
  }
  validators.validateIasOptions(iasConfig.options);
}

function isIasCheckRequired(appConfig) {
  return appConfig.authenticationMethod !== 'none' &&
    appConfig.routes.some(function(route) {
      return route.authenticationType === 'ias';
    });
}
