// @sap/approuter/lib/middleware/login-callback-middleware.js
'use strict';
const cookie = require('cookie');
const cookieUtils = require('../utils/cookie-utils');
const urlUtils = require('../utils/url-utils');
const headerUtil = require('../utils/header-util');
const oauthConfig = require('../passport/oauth-configuration.js');
const passport = require('passport');
const url = require('url');
const loginCallbackProvider = require('./login-callback-provider');
const VError = require('verror').VError;

module.exports = function passportLogin(req, res, next) {

  if (!loginCallbackProvider.isLoginCallback(req)) {
    return next();
  }

  let errorParam = urlUtils.getQueryParam(req, 'error');
  if (errorParam) {
    let errorDescriptionParam = urlUtils.getQueryParam(req, 'error_description');
    let errorFromQuery = new Error(errorDescriptionParam ? errorParam + ', ' + errorDescriptionParam : errorParam);
    return next(errorFromQuery);
  }
  if (process.env.PRESERVE_FRAGMENT !== 'false') {
    req.res = res;
  }
  oauthConfig.getXSUAAOauthStrategy(req, function (err, strategy) {
    if (err) {
      return next(err);
    }
    passport.use(strategy);
    passport.authenticate(strategy.name, function (err, user) {
      if (err) {
        let error = new VError(err, 'Could not authenticate with UAA');
        error.status = err.status;
        return next(error);
      }
      if (!user) {
        return sendRedirect(res, '/');
      }
      req.logIn(user,oauthConfig.SESSION_OPTIONS, function (err) {
        if (err) {
          return next(err);
        }
        const redirectCookieName = cookieUtils.getRedirectLocationCookieName();
        const fragmentCookieName = cookieUtils.getFragmentCookieName();

        if (!req.headers || !req.headers.cookie) {
          return sendRedirectCookieMissing(next, redirectCookieName);
        }

        const cookies = cookie.parse(req.headers.cookie);
        const redirectCookieValue = cookies[redirectCookieName];
        const fragmentCookieValue = cookies[fragmentCookieName];
        if (!redirectCookieValue) {
          return sendRedirectCookieMissing(next, redirectCookieName);
        }

        if (!isRelativeUrl(redirectCookieValue)) {
          return sendRedirectCookieError(next, 'Redirect path is invalid',
            new VError('%s must contain a relative path, got %s', redirectCookieName, redirectCookieValue));
        }
        let cookieOptions = {
          path: '/',
          maxAge: 0
        };
        let locationAfterLogin = cookie.serialize(redirectCookieName, '', cookieOptions);
        let samesite = req.routerConfig && req.routerConfig.cookies && req.routerConfig.cookies.SameSite;
        cookieUtils.setCookie(res, cookieUtils.addSamesite(locationAfterLogin, samesite));

        // Fragment may be empty string, which is casted to false. Therefore check presence of key.
        if (cookies.hasOwnProperty(fragmentCookieName)) {
          let fragmentAfterLogin = cookie.serialize(fragmentCookieName, '', cookieOptions);
          cookieUtils.setCookie(res, cookieUtils.addSamesite(fragmentAfterLogin,samesite));
        }

        if (cookies.hasOwnProperty('signature')) {
          let signature = cookie.serialize('signature', '', cookieOptions);
          cookieUtils.setCookie(res, cookieUtils.addSamesite(signature,samesite));
        }

        sendRedirect(res, normalizeRelativePath(redirectCookieValue + (fragmentCookieValue || '')));
      });
    })(req, res, next);
  });
};

function isRelativeUrl(location) {
  const parsedUrl = url.parse(location);
  return !(parsedUrl.host || parsedUrl.protocol);
}

function sendRedirectCookieMissing(next, cookieName) {
  sendRedirectCookieError(next, 'Missing redirect information', new VError('Required cookie \'%s\' is missing in the request', cookieName));
}

function sendRedirectCookieError(next, message, err) {
  const error = new VError(err, 'Unable to redirect after successful authentication. %s.', message);
  error.status = 400;
  next(error);
}

// normalize path, so it will not lead to unexpecred redirect like '//google.com' for example
function normalizeRelativePath(path) {
  return path.trim().replace(/^\/*/, '/');
}

function sendRedirect(res, redirectLocation) {
  res.writeHead(302, {
    'Location': redirectLocation,
    'Cache-Control': headerUtil.NOCACHE_HEADER_VALUE
  });
  res.end();
}
