// @sap/approuter/lib/passport/oauth-configuration.js
'use strict';

const _ = require('lodash');
const VError = require('verror');
const OAuth2Strategy = require('./oauth2-strategy');
const BasicOAuth2Strategy = require('./basic-oauth2-strategy');
const utils = require('./utils');

module.exports = {
  SESSION_OPTIONS : {
    keepSessionInfo: true
  },

  getUserSerializer: function (user, callback) {
    if (user && user.name) {
      callback(null, JSON.stringify(user));
    } else {
      callback(new Error('No user name found'));
    }
  },

  getUserDeserializer: function (sUser, callback) {
    if (sUser === '') {
      return callback(null, {});
    }
    let user;
    try {
      user = JSON.parse(sUser);
    } catch (err) {
      return callback(new VError(err, 'Could not parse user %s', sUser));
    }
    const error = user ? null : new VError('User not found %s', sUser);
    callback(error, user);
  },

  getXSUAAOauthStrategy: function (req, cb) {
    createStrategy(req, function (err, options, doneCallback) {
      if (err) {
        return cb(err);
      }
      cb(null, new OAuth2Strategy(options, doneCallback));
    });
  },

  getBasicOauthStrategy: function (req, credentials, cb) {
    createStrategy(req, function (err, options, doneCallback) {
      if (err) {
        return cb(err);
      }
      cb(null, new BasicOAuth2Strategy(_.assign(options, credentials), doneCallback));
    });
  }
};

function createStrategy(req, cb) {
  utils.loadOauthOptions(req, function (err, options) {
    if (err) {
      return cb(err);
    }

    function authenticationDone(req, params, done) {
      const tokenContext = {
        accessToken: params['access_token'],
        refreshToken: params['refresh_token'],
        expiresIn: params['expires_in'],
        scope: params.scope,
        xsuaaToken: params['xsuaaToken'],
        urlTenant: params['urlTenant'],
        idToken: params['id_token'],
        oauthOptions: options
      };

      utils.storeToken(req, tokenContext, function (err) {
        if (!err) {
          req.app.approuter.emit('login', req.session);
        }
        done.apply(null, arguments);
      });
    }

    cb(null, options, authenticationDone);
  });
}
